numba/cuda/reduction.py

```diff
@@ -0,0 +1,296 @@
+"""
+A library written in CUDA Python for generating reduction kernels
+"""
+from __future__ import print_function, division, absolute_import
+from functools import reduce
+import math
+from numba.numpy_support import from_dtype
+from numba.types import uint32
+
+def reduction_template(binop, typ, blocksize):
+    """
+    Args
+    ----
+    binop : function object
+        A binary function as the reduction operation
+    typ : numba type
+        The numba type to the reduction operation
+    blocksize : int
+        The CUDA block size (thread per block)
+    """
+    from numba import cuda
+    from numba import intp
+
+    if blocksize > 512:
+        # The reducer implementation is limited to 512 threads per block
+        raise ValueError("blocksize too big")
+
+    # Compile binary operation as device function
+    binop = cuda.jit((typ, typ), device=True)(binop)
+
+    # Compile reducer
+    @cuda.jit((typ[:], typ[:], intp, intp))
+    def reducer(inp, out, nelem, ostride):
+        tid = cuda.threadIdx.x
+        i = cuda.blockIdx.x * (blocksize * 2) + tid
+        gridSize = blocksize * 2 * cuda.gridDim.x
+        sdata = cuda.shared.array(blocksize, dtype=typ)
+
+        while i < nelem:
+            sdata[tid] = binop(inp[i], inp[i + blocksize])
+            i += gridSize
+
+        cuda.syncthreads()
+
+        if blocksize >= 512:
+            if tid < 256:
+                sdata[tid] = binop(sdata[tid], sdata[tid + 256])
+                cuda.syncthreads()
+
+        if blocksize >= 256:
+            if tid < 128:
+                sdata[tid] = binop(sdata[tid], sdata[tid + 128])
+                cuda.syncthreads()
+
+        if blocksize >= 128:
+            if tid < 64:
+                sdata[tid] = binop(sdata[tid], sdata[tid + 64])
+                cuda.syncthreads()
+
+        if tid < 32:
```

Comment:
Is there a rationale for how the code is organized here? Above, the tid checks are done into each blocksize-specific block, but here it's only done once. It would be nice to comment on that.